core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -554,25 +477,34 @@ private object RTS {
                     curIo = enterSupervision *>
                       io.value.ensuring(exitSupervision(io.supervisor))
 
-                  case IO.Tags.Terminate =>
-                    val io = curIo.asInstanceOf[IO.Terminate]
+                  case IO.Tags.Fail =>
+                    val io = curIo.asInstanceOf[IO.Fail[E]]
 
-                    val finalizer = interruptStack
+                    val finalizer = unwindStack(io.cause.isChecked)
```

Comment:
Right now it's only catching the exception if the cause is checked. I'm not sure that's correct because `Redeem` now—quite intentionally—lets us "catch" unchecked and interruption. We are not using that, I don't think, but if we're doing the correct thing in "lifting" the old methods (redeem, etc.), up into `Redeem`, then we won't need this logic here. So I _think_ you want to delete the boolean to `interruptStack` and always let the catcher catch the exception.

In a later pull request, I'll add some "hints" that let us know when the cause will be caught and used, versus caught and then rethrown. This will be strictly for performance, though, and will not be required for correctness.